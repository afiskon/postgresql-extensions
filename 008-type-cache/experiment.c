#include "postgres_ext.h"
#include <postgres.h>
#include <utils/builtins.h>
#include <utils/typcache.h>

PG_MODULE_MAGIC;
PG_FUNCTION_INFO_V1(experiment_max);

Datum
experiment_max(PG_FUNCTION_ARGS)
{
    Datum a = PG_GETARG_DATUM(0);
    Datum b = PG_GETARG_DATUM(1);
    Oid collation = PG_GET_COLLATION();
    Oid elmtyp = get_fn_expr_argtype(fcinfo->flinfo, 0);
    TypeCacheEntry *typentry;
    int cmp_result;

    typentry = (TypeCacheEntry *) fcinfo->flinfo->fn_extra;
    if (typentry == NULL || typentry->type_id != elmtyp)
    {
        typentry = lookup_type_cache(elmtyp, TYPECACHE_CMP_PROC_FINFO);
        fcinfo->flinfo->fn_extra = (void *) typentry;
    }

    cmp_result = DatumGetInt32(FunctionCall2Coll(&typentry->cmp_proc_finfo, collation, a, b));
    return cmp_result <= 0 ? b : a;
}
