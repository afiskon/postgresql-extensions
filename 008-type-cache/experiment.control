comment = 'Experiment: yet another experiment with PostgreSQL extensions'
default_version = '1.0'
module_pathname = '$libdir/experiment'
relocatable = false
