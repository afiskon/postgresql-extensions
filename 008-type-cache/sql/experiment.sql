CREATE EXTENSION experiment;

SELECT experiment_max(1, 2);
SELECT experiment_max(true, false);
SELECT experiment_max('aaa', 'bbb' :: text COLLATE "C");
SELECT experiment_max('{1,2,3}', '{4,5,6}' :: int[]);

DROP EXTENSION experiment;
