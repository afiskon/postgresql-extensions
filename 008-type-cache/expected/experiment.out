CREATE EXTENSION experiment;
SELECT experiment_max(1, 2);
 experiment_max 
----------------
              2
(1 row)

SELECT experiment_max(true, false);
 experiment_max 
----------------
 t
(1 row)

SELECT experiment_max('aaa', 'bbb' :: text COLLATE "C");
 experiment_max 
----------------
 bbb
(1 row)

SELECT experiment_max('{1,2,3}', '{4,5,6}' :: int[]);
 experiment_max 
----------------
 {4,5,6}
(1 row)

DROP EXTENSION experiment;
